testy.go

```diff
@@ -0,0 +1,133 @@
+)
+
+type testy struct {
+	tests orderedmap.OrderedMap[string, orderedmap.OrderedMap[string, *Test]]
+}
+
+var instance testy
+
+// TestingT is a subset of testing.T that we have to implement for non-`go test` runs.
+//
+// TODO flesh this out with more useful stuff from testing.T
+type TestingT interface {
+	Fail()
+	FailNow()
+	Fatal(args ...interface{})
+	Fatalf(format string, args ...interface{})
+	Errorf(format string, args ...interface{})
+	Helper()
+	Log(args ...interface{})
+	Logf(format string, args ...interface{})
+	Name() string
+}
+
+type Tester func(t TestingT)
+
+type Test struct {
+	Package string
+	Name    string
+	tester  Tester
+	// TODO better results
+	Msgs   []string
+	Failed bool
+}
+
+var regLock sync.Mutex
+
+func RegisterTest(name string, tester Tester) interface{} {
+	// we only care about our immediate caller
+	callers := make([]uintptr, 1)
+	// skip over Callers and ourselves
+	n := runtime.Callers(2, callers[:])
+	pkg := "<unknown>"
+	// make sure it was able to actually get our caller
+	if n > 0 {
+		frames := runtime.CallersFrames(callers)
+		frame, _ := frames.Next()
+		// remove the function name (which is almost certainly "init") and leave just the package name.
+		// as an example, this function is `github.com/gametimesf/testy.RegisterTest`.
+		i := strings.LastIndex(frame.Function, ".")
+		pkg = frame.Function[:i]
+	}
+
+	regLock.Lock()
+	defer regLock.Unlock()
+	if instance.tests == nil {
+		instance.tests = make(orderedmap.OrderedMap[string, orderedmap.OrderedMap[string, *Test]])
+	}
+	if instance.tests[pkg] == nil {
+		instance.tests[pkg] = make(orderedmap.OrderedMap[string, *Test])
+	}
+
+	if _, exists := instance.tests[pkg][name]; exists {
+		panic(fmt.Sprintf("test %s already exists in package %s", name, pkg))
+	}
+
+	instance.tests[pkg][name] = &Test{
+		Package: pkg,
+		Name:    name,
+		tester:  tester,
+	}
+
+	return nil
+}
+
+// RunAsTest runs all registered tests under Go's testing framework. To run tests on a per-package basis, put a test
+// file in each package containing a single test that calls this function. This is recommended so accurate per-package
+// execution times are reported, as well as using the test cache. Do not import a test package into another test package
+// as that will cause the tests in the second package to get executed with the first package. If code or resources need
+// shared between test packages, put them in their own package which does not contain any test definitions.
+func RunAsTest(t *testing.T) {
+	t.Helper()
+	for _, tests := range instance.tests {
+		for _, test := range tests {
+			t.Run(test.Name, func(tt *testing.T) {
+				tt.Helper()
+				test.tester(tt)
+			})
+		}
+	}
+}
+
+// Run runs all registered tests and returns result information about them.
+//
+// TODO: ability to filter for specific packages and tests
+// TODO: channel for results to support progressive progress loading? or some sort of ID and background processing?
```

Comment:
+1 for background processing, or just running these concurrently in a single wait group.

Reply:
running everything concurrently is not going to end well since there will be interdependencies. it should be somewhat low-hanging to run each package concurrently (like `go test` does by default), but supporting `t.Parallel` to run individual tests in parallel looks somewhat nontrivial (and we usually don't actually write parallel tests anyway)